cat/opencl.cat

```diff
@@ -60,7 +60,7 @@ let ra_sw(r) = ((r & [Rel]); Fsb?; [W \ WI]; rs?; r; rf; [R \ WI]; sbF?; ([Acq]
 let bar_sw(r) = (r & [EF]); syncbar & ~wi & swg; (r & [XF])
 
 (* Allowed to synchronise on the other region *)
-let scf = (SC * SC) | ((dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)); F*F;(dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)))
+let scf = (SC * SC) | ((dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)); unv; (dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)))
```

Comment:
you don't need to `unv`, do you?

Reply:
I think we need? Otherwise it becomes `([DG] & [DL]);([DG] & [DL])` which is `([DG] & [DL] & [DG] & [DL])`

Reply:
[x];unv;[y] should give you all pairs in unv (which is pretty much anything) restricted to pairs starting with x and ending with y. I fail to see how this is different than [x];[y]

Reply:
> I think we need? Otherwise it becomes `([DG] & [DL]);([DG] & [DL])` which is `([DG] & [DL] & [DG] & [DL])`

I don't this this is true. The first one allows you to have pairs (x,y) with x!=y, the second ones does not

Reply:
`unv` is defined as `_ * _`, so I think `[X]; unv; [Y]` should be `X * Y`? And `[X];[Y]` is like `X & Y`

Reply:
> `unv` is defined as `_ * _`, so I think `[X]; unv; [Y]` should be `X * Y`?

This I agree

> And `[X];[Y]` is like `X & Y`

This I don't. The first uses composition so it is clearly a relation. The second one is the intersection of tags, thus, it is a tag

Reply:
I see that the change I propose makes `overhauling/example6` fail, but I suspect that the problem is that our AtomiceFence event misses the VISIBLE tag

Reply:
`[X];[Y] == [X & Y] ~ X & Y` is true, because you compose identity relations. What Haining does is correct.

Reply:
Does this mean that we need `unv`, otherwise the result relation would be a subset of `id`?

Reply:
Yes. Composing identities gives you only identities. If the intention is to have a cartesian-product-like relation, then you need to have a proper binary relation somewhere (either `F*F` as in the original or `unv` like now).